tests/ops/op_math/test_controlled.py

```diff
@@ -781,63 +796,228 @@ def test_sparse_matrix_format(self):
+        [1],
+        [0],
+        qml.CRX,
+        [
+            qml.RZ(np.pi / 2, wires=1),
+            qml.RY(0.123 / 2, wires=1),
+            qml.CNOT(wires=[0, 1]),
+            qml.RY(-0.123 / 2, wires=1),
+            qml.CNOT(wires=[0, 1]),
+            qml.RZ(-np.pi / 2, wires=1),
+        ],
+    ),
+    (
+        qml.RY,
+        [0.123],
+        [1],
+        [0],
+        qml.CRY,
+        [
+            qml.RY(0.123 / 2, 1),
+            qml.CNOT(wires=(0, 1)),
+            qml.RY(-0.123 / 2, 1),
+            qml.CNOT(wires=(0, 1)),
+        ],
+    ),
+    (
+        qml.RZ,
+        [0.123],
+        [0],
+        [1],
+        qml.CRZ,
+        [
+            qml.PhaseShift(0.123 / 2, wires=0),
+            qml.CNOT(wires=[1, 0]),
+            qml.PhaseShift(-0.123 / 2, wires=0),
+            qml.CNOT(wires=[1, 0]),
+        ],
+    ),
+    (
+        qml.Rot,
+        [0.1, 0.2, 0.3],
+        [1],
+        [0],
+        qml.CRot,
+        [
+            qml.RZ((0.1 - 0.3) / 2, wires=1),
+            qml.CNOT(wires=[0, 1]),
+            qml.RZ(-(0.1 + 0.3) / 2, wires=1),
+            qml.RY(-0.2 / 2, wires=1),
+            qml.CNOT(wires=[0, 1]),
+            qml.RY(0.2 / 2, wires=1),
+            qml.RZ(0.3, wires=1),
+        ],
+    ),
+    (
+        qml.PhaseShift,
+        [0.123],
+        [1],
+        [0],
+        qml.ControlledPhaseShift,
+        [
+            qml.PhaseShift(0.123 / 2, wires=0),
+            qml.CNOT(wires=[0, 1]),
+            qml.PhaseShift(-0.123 / 2, wires=1),
+            qml.CNOT(wires=[0, 1]),
+            qml.PhaseShift(0.123 / 2, wires=1),
+        ],
+    ),
+]
+
+custom_ctrl_op_decomps = special_non_par_op_decomps + special_par_op_decomps
+
+pauli_x_based_op_decomps = [
+    (qml.PauliX, [0], [1], [qml.CNOT([1, 0])]),
+    (
+        qml.PauliX,
+        [2],
+        [0, 1],
+        qml.Toffoli.compute_decomposition(wires=[0, 1, 2]),
+    ),
+    (
+        qml.CNOT,
+        [1, 2],
+        [0],
+        qml.Toffoli.compute_decomposition(wires=[0, 1, 2]),
+    ),
+    (
+        qml.PauliX,
+        [3],
+        [0, 1, 2],
+        qml.MultiControlledX.compute_decomposition(wires=[0, 1, 2, 3], work_wires=Wires("aux")),
+    ),
+    (
+        qml.CNOT,
+        [2, 3],
+        [0, 1],
+        qml.MultiControlledX.compute_decomposition(wires=[0, 1, 2, 3], work_wires=Wires("aux")),
+    ),
+    (
+        qml.Toffoli,
+        [1, 2, 3],
+        [0],
+        qml.MultiControlledX.compute_decomposition(wires=[0, 1, 2, 3], work_wires=Wires("aux")),
+    ),
+]
+
+
+class TestDecomposition:
+    """Test decomposition of Controlled."""
+
+    @pytest.mark.parametrize(
+        "target, decomp",
+        [
+            (
+                OpWithDecomposition(0.123, wires=[0, 1]),
+                [
+                    qml.CH(wires=[2, 0]),
+                    Controlled(qml.S(wires=1), control_wires=2),
+                    qml.CRX(0.123, wires=[2, 0]),
+                ],
```

Comment:
Nice, this is equivalent to the test we talked about in https://github.com/PennyLaneAI/pennylane/pull/5113#discussion_r1468208850 right?

Reply:
Yes.